design/sales.md

```diff
@@ -45,3 +45,49 @@ the Reservations module, so that it survives node restarts.
     Datastore API:
       function put(key, value)
       function get(key): value
+
+Reserving storage space
+-----------------------
+
+When a request for storage is submitted on chain, the sales module decides
+whether or not it wants to act on it. If the requested duration and size match
+what the node wants to sell and the reward is sufficient, then the sales module
+will go through several steps to try and fill a slot in the request.
+
+First, it will select a slot from the request to fill. Then, it will reserve
+space in the Repo to ensure that it keeps enough space available to store the
+content. It will mark the reserved space as belonging to the slot. Next, it will
+download the content, calculate a storage proof, and submit the proof on chain.
+If any of these later steps fail, then the node should release the storage that
+it reserved earlier.
+
+When a slot was filled successfully, then its storage should not be released
+until the request ends.
+
+Releasing storage space
+-----------------------
+
+When a storage request ends, or an attempt to fill a slot fails, it is important
+to release the reserved space that belonged to the slot. To ensure that the
+right amount of space is released, and that it is only released once, we keep
+track of how much space is reserved in the Repo for a slot.
+
+Releasing storage space goes in three steps. First, we look up the amount of
+space that is reserved for the slot in the Datastore. Then we release that
+amount in the Repo, and remove the entry for the slot from the Datastore.
+
+The first step ensures that we release the correct amount of space from the
+Repo. The last step ensures that we do not release space from the Repo more than
+once.
+
+Persisting state
+----------------
+
+The sales module keeps state in a number of places. Most state is kept on chain,
+this includes the slots that a host is filling and the state of each slot. This
+ensures that a node's local view of slot states does not deviate from the
+network view, even when the network changes while the node is down. The rest of
+the state is kept on local disk by the Repo and the Datastore. How much space is
+reserved by the sales module, and how much remains available to sell is
+persisted on disk by the Repo. A mapping between amounts of reserved space and
```

Comment:
> How much space is reserved by the sales module, and how much remains available to sell is
persisted on disk by the Repo

How can we distinguish which bytes are reserved for the sales module and which bytes were reserved for other parts of the node? Perhaps we should store some of this information in the metadata datastore?

Reply:
Actually, I think this was already covered above:
> To ensure that the right amount of space is released, and that it is only released once, we keep
track of how much space is reserved in the Repo for a slot.

Maybe we should specify it as being tracked as metadata?

Reply:
> How can we distinguish which bytes are reserved for the sales module and which bytes were reserved for other parts of the node? Perhaps we should store some of this information in the metadata datastore?

Is this needed -  not really. The repo only cares about how much is used vs reserved vs available, wether it was set by module A or module B, really doesn't matter from the repo perspective, nor should it from the system perspective as a whole.

> track of how much space is reserved in the Repo for a slot.

We only need to keep track of this while in the process of filling the slot (i.e. downloading data) and we only care about what blocks to delete if something goes wrong, other than that, the repo already keeps track of usages on every write - once a block is written on disk, the repo counts this as used, if it's deleted, it counts that as free. Also, if a slot was filled correctly, the repo will make sure to cleanup the blocks automatically after the slots expires.

Reply:
> the repo will make sure to cleanup the blocks automatically after the slots expires.

How will the repo know to cleanup expired slots?

Reply:
> > the repo will make sure to cleanup the blocks automatically after the slots expires.
> 
> How will the repo know to cleanup expired slots?

Each block has an expiration ttl, once a block expires it's picked up by the manager module and cleaned up.

Reply:
I see, so we'll need to specify the request/slot expiry in the `putBlock` calls when downloading. Once a request has started, we need to extend the ttl of those blocks to be the storage request duration.

Reply:
> > How can we distinguish which bytes are reserved for the sales module and which bytes were reserved for other parts of the node? Perhaps we should store some of this information in the metadata datastore?
>
> Is this needed - not really. The repo only cares about how much is used vs reserved vs available, wether it was set by module A or module B, really doesn't matter from the repo perspective, nor should it from the system perspective as a whole.

When an availability is created, we would need to check if the repo has already enough reservation to accomodate that availability. If not, reserve more. How will we know we are checking reserved space for the sales module and not reserved space for some other module? For example, ModuleA reserves 1000MiB, and SalesModule reserves 500MiB (initially as the "large chunk"). Querying the repo, we get a reserved total space of 1500MiB. An availability of 500MiB is created, which logically consumes all of the 500MiB of reserved SalesModule space. Querying the repo again, we get a reserved total space of 1000MiB. If another availability of 500MiB is created, it would appear as if the repo already has enough space reserved (1000MiB) and that no additional space would need to be reserved. Shouldn't there be a way to understand which module has reserved which amount of space, or is that something we should keep track of in the metadata datastore?

Reply:
We don't foresee any other module reserving storage, only the sales module ever needs to reserve storage, the total reserved storage should match the total storage of all the availabilities.

Reply:
> We don't foresee any other module reserving storage, only the sales module ever needs to reserve storage, the total reserved storage should match the total storage of all the availabilities.

If possibly in the future, a user wanted to set aside X bytes of data to be a caching node (and take advantage of bandwidth incentives), would they use this reservation api? Or is something else being considered?

Reply:
After the large chunk is initially reserved, let's say that all that storage is eventually released by completed contracts. The next time an availability comes in, what should happen? Should the size specified in the availability be reserved? Or should the large chunk (% of user's disk) again be reserved?

Reply:
> If possibly in the future, a user wanted to set aside X bytes of data to be a caching node (and take advantage of bandwidth incentives), would they use this reservation api? Or is something else being considered?

The notion of reserving space is specific to the renting model (and module), where you want to set some space aside without actually using it. I don't see how a caching node would use this and right now, but if the need arrises, the reservation api could certainly be used for that as well. I would not however try to go after this hypothetical scenario right now, even if we end up requiring it, it will be sometime in the future, right now bandwidth is a future feature.

Reply:
Would we want to consider the "large chunk" CLI param to be "the maximum amount of storage to rent out"?